react/OrderForm.tsx

```diff
@@ -68,30 +145,42 @@ export const OrderFormProvider: FC = ({ children }) => {
       console.error(error.message)
     }
 
-    if (loadingQuery) {
+    const localOrderFormString = localStorage.getItem('orderform')
+
+    if (localOrderFormString != null) {
+      const localOrderForm = JSON.parse(localOrderFormString)
+
+      if (localOrderForm.value !== UNSYNC_ORDER_FORM_VALUE) {
+        return
+      }
+    }
+
+    if (loading || error || !data) {
       return
     }
 
-    setLoading(false)
+    updateOrderFormCache(data.orderForm)
+    setOrderForm(data.orderForm)
+  }, [data, error, loading, updateOrderFormCache])
 
-    if (data) {
-      setOrderForm(data.orderForm)
-      updateApolloCache(client, data.orderForm)
-    }
-  }, [client, data, error, loadingQuery])
+  useEffect(() => {
+    saveLocalOrderForm(orderForm)
+  }, [orderForm])
 
-  const value = useMemo(
+  const value = useMemo<Context>(
     () => ({
       error,
-      loading,
-      orderForm: error ? emptyOrderForm : orderForm,
-      setOrderForm: (newOrderForm: Partial<OrderForm>) => {
-        updateApolloCache(client, { ...orderForm, ...newOrderForm })
-        setOrderForm(newOrderForm)
+      orderForm: {
+        ...orderForm,
+        value:
+          orderForm.value === UNSYNC_ORDER_FORM_VALUE ? 0 : orderForm.value,
       },
+      setOrderForm,
```

Comment:
I think so, do you think it would work if I updated inside a `useEffect`? so I will only need to update the cache in a single place

Reply:
I think it would, that would be even better than the previous implementation. It would need some testing though, since not updating the cache causes inconsistencies when navigating between pages.

Reply:
weird enough, everytime I add a new item to the order form I get a bunch of warnings on the console for missing fields on the item (probably because they exist in the orderForm query but no on the "local" item I'm adding) such as `__typename`.

I'm thinking if there is a way to only update the order form cache when the value passed to `setOrderForm` is from the server, and not a local update

Reply:
yeah, I won't be able to use it in a `useEffect`, because apollo throws when you pass an object to `writeQuery` that doesn't have `__typename`s, and since we are doing offline-first operations (and they won't have the typename until its sent to the server) it's always going to break.

I couldn't think of a good enough solution that would cover this, but I'm thinking what are the consequences of not updating the cache everytime the local order form changes. I haven't seen any inconsistencies when changing pages but maybe you know a scenario that is garanteed to not work?

Reply:
That was a consistent issue that happened iirc when you added an item to cart then navigated to a different department, for instance. The minicart icon would display the incorrect amount of items, because when the component loaded for the first time, it would query for the orderform and use the value in cache, which was outdated. It doesn't seem to reproduce in your workspace, though.
I'll take a deeper look at this code later to understand what's going on. Maybe it won't be necessary to update the cache after all.

Reply:
nice! let me know if you need any assistance in understanding the code :-)